GSoC/2016/JuliaQuantum-:-Framework-for-QuComput.md

```diff
@@ -0,0 +1,420 @@
+    + **graph states** graph states visualization will be based on [GraphPlot.jl](https://github.com/afternone/GraphPlot.jl)
+
+At the end of the project, JuliaQuantum will be equipped with a framework to simulate quantum computation. Including the features about:
+
+- quantum circuit's simulation and visualization
+- Adiabatic model simulators
+- one-way model's simulation and visualization
+
+The project will make use of existing packages and aims to contribute back, for example:
+
+- `QuArray` will be used to store the states of Qubits
+- QuDynamics.jl will provide adiabatic computational model with more efficient Shordinger equation solvers with good accuracy. 
+
+
+Explanation of nouns:
+
+- **computational models** here means the algorithms that can provide universal computation and is possible to be realized in experiments
+
+
+---
+## Details :
+
+### Structure of Types
+
+![](imag/AbstractGate.png)
+
+
+### Prototype
+I wrote a prototype for my idea with only simple features about quantum circuit and adiabatic quantum simulation module in [Kebab](https://github.com/Roger-luo/Kebab).
+
+### Implementation:
+
+#### Type System
+- **Quantum Gates** `AbstractGates` is a super type for all modules in the structure of this framework. The simulations of a quantum computation model will be construct in a quantum-circuit-like way. The other quantum computation model will be a gate-like module(subtype of `AbstractGates`) in quantum circuit.
+- **Quantum Gate** quantum gates `Gate` includes gate-like modules and single qubit gates like Hadamard gate,Pauli gates...
+- **Graph** graph type will be used to store graph states for one-way model, the type is current considered to use adjacent matrix and use [GraphPlot.jl](https://github.com/afternone/GraphPlot.jl) as dependency to visualize
+- **Computational Models** `QuComputModel` is the the basic type for other models which is a subtype of AbstractGates
+
+At the end of the project, JuliaQuantum will be equipped with a framework to simulate quantum computation models including quantum circuits, adiabatic quantum computation, one-way model, and quantum error corrections with CSS code, graph code.
+
+
+---
+### Design :
+
+The idea is to organize solvers, quantum gates and other quantum computation simulators in the form of quantum gates.
+
+```julia
+abstract AbstractGates
+abstract QuCircuit
+
+# Basic Types for states
+## should QuBit become a subtype of QuArray? or I just write my own type first?
+
+type QuBit #or AbstractVector
+    state::AbstractVector
+    bitnum::Integer
+
+    function QuBit(state::AbstractVector,bitnum::Integer)
+        check_state(state) #check if state length meets bitnum
+        new(state,bitnum)
+    end
+end
+
+## A graph state is stored by an adjacent matrix
+type GraphState
+    adjaMatrix::AbstractMatrix
+end
+
```

Comment:
Ok, I need to attend a experiment course this afternoon, will update a new structure of prototype when I'm back in 2 hours. :-)